sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs

```diff
@@ -1,40 +1,98 @@
+                select p;
+            
+            package = matches.FirstOrDefault();
+            return package != null;
         }
 
-        internal static bool TryGetResourcePackage(string name, string version, [NotNullWhen(true)] out IResourcePackage? package)
+        private static ImmutableList<IResourcePackage> DiscoverResourcePackages()
         {
-            return _resourcePackages.TryGetValue(PackageKey(name, version), out package);
+            return LoadReferencedAssemblies()
+                .SelectMany(s => s.GetTypes())
+                .Where(typeof(IResourcePackage).IsAssignableFrom)
+                .SelectMany(type => type.GetConstructors().Where(c => c.GetParameters().Length == 0))
+                .Select(c => c.Invoke(new object[0]))
+                .Cast<IResourcePackage>()
+                .ToImmutableList();
         }
-
-        public static void RegisterResourcePackage(string name, string version, IResourcePackage package)
+        
+        // Assemblies are loaded on demand, so it could be that some assemblies aren't yet loaded to the current
+        // app domain at the time of discovery. This method iterates through the list of referenced assemblies
+        // recursively.
+        // Note: If an assembly is referenced but its types are never used anywhere in the program, that reference
+        // will be optimized out and won't appear in the result of the enumeration.
+        private static IEnumerable<Assembly> LoadReferencedAssemblies()
         {
-            var key = PackageKey(name, version);
-            if (!_resourcePackages.TryAdd(key, package))
+            var yieldedAssemblies = new HashSet<string>();
+            var assembliesToCheck = new Queue<Assembly>();
+
+            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
             {
-                throw new InvalidOperationException($"Cannot re-register package {key}.");
+                if (PossibleMatch(assembly.GetName()))
+                {
+                    assembliesToCheck.Enqueue(assembly!);
+                }
             }
+
+            while (assembliesToCheck.Any())
+            {
+                Assembly assemblyToCheck = assembliesToCheck.Dequeue();
+                if (yieldedAssemblies.Contains(assemblyToCheck.FullName!))
+                    continue;
+
+                yieldedAssemblies.Add(assemblyToCheck.FullName!);
+                yield return assemblyToCheck;
+                
+                foreach (var reference in assemblyToCheck.GetReferencedAssemblies())
+                {
+                    if (PossibleMatch(reference))
+                    {
+                        var assembly = Assembly.Load(reference);
+                        assembliesToCheck.Enqueue(assembly);
+                    }
+                }
+            }
+
+            static bool PossibleMatch(AssemblyName? assembly) => assembly != null && !assembly.FullName.StartsWith("System");
```

Comment:
Nit: `StartsWith` does a culture-sensitive comparison by default. Best to specify an explicit `StringComparison`.

```suggestion
            static bool PossibleMatch(AssemblyName? assembly) => assembly != null && !assembly.FullName.StartsWith("System", StringComparison.Ordinal);
```